### PR TITLE
Fix Ogre2 assertions during ray queries

### PR DESCRIPTION
**Pull request by darksylinc**
## Summary

This bug had been annoying me for months for which I silenced those asserts.

But given that I need to implement `SCENE_STATIC` to get global illumination going, it is dangerous to ignore those asserts so I had to fix this.

ign-rendering would assert (inside a debug build of Ogre2) every time the mouse hovers the GUI.

This PR changes both API and ABI. Technically speaking what ign-rendering was doing was wrong but it didn't seem to matter too much (if at all); which is why I don't care to break API/ABI without backporting.

If this fix makes it into Fortress then great, otherwise it's not a big deal. Ideally ogre2 fork should include [this fix](https://github.com/OGRECave/ogre-next/commit/e7f3d114102e125ab599346aa3dd4b722a9fac07) we added to 2.3 and that I just backported to 2.2; but again no rush.

## codecheck

Complains:

> /home/matias/Projects/ign/src/ign-rendering/include/ignition/rendering/base/BaseRayQuery.hh:146: Using C-style cast.  Use static_cast<bool>(...) instead  [readability/casting] [4]

It seems it doesn't like the argument getting commented out:

```cpp
RayQueryResult BaseRayQuery<T>::ClosestPoint(bool /*_forceSceneUpdate*/) {} // complains
RayQueryResult BaseRayQuery<T>::ClosestPoint(bool _forceSceneUpdate) {} // does not complain
```

However doing so will cause `unused argument` warning during build.
It may be possible to silence both using an UNUSED() macro if it doesn't exist already:

```cpp
#define IGN_UNUSED(x) do { (void)sizeof(x); } while(0)

RayQueryResult BaseRayQuery<T>::ClosestPoint(bool _forceSceneUpdate)
{
  IGN_UNUSED(_forceSceneUpdate);
}

// Alternatively C++17
RayQueryResult BaseRayQuery<T>::ClosestPoint([[maybe_unused]] bool _forceSceneUpdate) {}

// Attributes on GCC/Clang
RayQueryResult BaseRayQuery<T>::ClosestPoint(__attribute__((unused)) bool _forceSceneUpdate) {}

// It may be best to use a macro to abstract these changes:
RayQueryResult BaseRayQuery<T>::ClosestPoint(IGN_UNUSED_ARG bool _forceSceneUpdate) {}
```

## Commit msg

This change negatively affects performance as we now force a scene
update with each query.

Users can specify ClosestPoint(false) if they need to call this function
multiple times without modifying anything or if they know the scene is
up to date

Signed-off-by: Matias N. Goldberg <dark_sylinc@yahoo.com.ar>

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

**Review by ahcorde — CHANGES_REQUESTED** [review]
```
  /github/workspace/include/ignition/rendering/base/BaseRayQuery.hh:146:  Using C-style cast.  Use static_cast<bool>(...) instead  [readability/casting] [4]
  /github/workspace/ogre/src/OgreRayQuery.cc:63:  Using C-style cast.  Use static_cast<bool>(...) instead  [readability/casting] [4]
```

**Comment by darksylinc** [comment]
> ```
>   /github/workspace/include/ignition/rendering/base/BaseRayQuery.hh:146:  Using C-style cast.  Use static_cast<bool>(...) instead  [readability/casting] [4]
>   /github/workspace/ogre/src/OgreRayQuery.cc:63:  Using C-style cast.  Use static_cast<bool>(...) instead  [readability/casting] [4]
> ```

Hi please read the ticket above. I need guidance on which solution you want to fix this false positive from codecheck

File: include/ignition/rendering/base/BaseRayQuery.hh
Comment (ahcorde) [inline include/ignition/rendering/base/BaseRayQuery.hh:146]:
```suggestion
    RayQueryResult BaseRayQuery<T>::ClosestPoint(bool /*_forceSceneUpdate*/)  // NOLINT
```

**Review by ahcorde — CHANGES_REQUESTED** [review]
> > ```
> >   /github/workspace/include/ignition/rendering/base/BaseRayQuery.hh:146:  Using C-style cast.  Use static_cast<bool>(...) instead  [readability/casting] [4]
> >   /github/workspace/ogre/src/OgreRayQuery.cc:63:  Using C-style cast.  Use static_cast<bool>(...) instead  [readability/casting] [4]
> > ```
> 
> Hi please read the post above. I need guidance on which solution you want to fix this false positive from codecheck

I would say that adding `// NOLINT` will fix this false positives

**Review by chapulina — APPROVED** [review]
LGTM, we just need to make sure the linters are happy before merging.

**Comment by darksylinc** [comment]
> LGTM, we just need to make sure the linters are happy before merging.

I haven't got the time yet to fix it, but I haven't forgotten about this PR. Sorry it's taking so long.

File: include/ignition/rendering/base/BaseRayQuery.hh
Comment (darksylinc) [reply include/ignition/rendering/base/BaseRayQuery.hh]: Suggested fix applied

**Comment by darksylinc** [comment]
OK I've rebased and updated the PR to latest main.

Now we're waiting on the CI. Crossing fingers it works out :crossed_fingers:

**Review by ahcorde — APPROVED** [review]

**Pull request merged**
